src/messages.rs

```diff
@@ -0,0 +1,299 @@
+            return Err(io::Error::new(io::ErrorKind::Other, "Wrong magic bytes"));
+        }
+        let (header, buf) = match try!(Header::from_bytes(buf)) {
+            None => return Ok(None),
+            Some((header, buf)) => (header, buf),
+        };
+        let (payload, buf) = match try!(Payload::from_bytes(buf, secp256k1)) {
+            None => return Ok(None),
+            Some((payload, buf)) => (payload, buf),
+        };
+        return Ok(Some((
+            Message {
+                header: header,
+                payload: payload,
+            },
+            buf,
+        )));
+    }
+    fn to_bytes(&self, buf: &mut BytesMut, secp256k1: &Secp256k1) -> io::Result<()> {
+        buf.extend(vec![0x68, 0x1e, 0x58, 0x12]);
+        self.header.to_bytes(buf)?;
+        self.payload.to_bytes(buf, secp256k1)?;
+        Ok(())
+    }
+}
+
+/// Headers take the following form:
+///
+/// | Field   | Length | Description                                                   |
+/// |---------|--------|---------------------------------------------------------------|
+/// | Version | 1      | Protocol version                                              |
+///
+#[derive(Copy, Clone, PartialEq, Eq, Debug)]
+struct Header {
+    protocol_version: u8,
+}
+
+impl Header {
+    fn new(protocol_version: u8) -> Header {
+        Header { protocol_version: protocol_version }
+    }
+    /// Takes a byte slice and builds a Header. Returns same as MessageCodec::decode plus the
+    /// remaining (unconsumed) buffer.
+    fn from_bytes<'a>(buf: &'a [u8]) -> io::Result<Option<(Header, &'a [u8])>> {
+        if buf.len() < 1 {
+            return Ok(None);
+        }
+        let (ver, buf) = buf.split_at(1);
+        return Ok(Some((Header { protocol_version: ver[0] }, buf)));
+    }
+    fn to_bytes(&self, buf: &mut BytesMut) -> io::Result<()> {
+        buf.extend(vec![self.protocol_version]);
+        Ok(())
+    }
+}
+
+/// Payloads take the following form:
+///
+/// | Field   | Length | Description                                                   |
+/// |---------|--------|---------------------------------------------------------------|
+/// | Variant | 1      | Determines which payload variant this is                      |
+/// | Variant | ...    | The actual variant                                            |
+///
+#[derive(Copy, Clone, PartialEq, Eq, Debug)]
+enum Payload {
+    /// KeyExchange take the following form:
+    ///
+    /// | Field      | Length | Description                                                        |
+    /// |------------|--------|--------------------------------------------------------------------|
+    /// | Signature  | 64     | Compressed ECDSA signature over the following public key using the |
+    /// |            |        | long term  public key                                              |
+    /// | Public Key | 33     | Compressed ephemeral public key                                    |
```

Comment:
Okay, there won't be too many public keys. Still would be good to call that struct field `ke_pk` (for key exchange public key)